File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        # iterate from left to right to store prefix of each
        prefix = postfix = 1
        
        ans = []
        
        for i, num in enumerate(nums):
            if i == 0:
                ans.append(1)
            else:
                prefix *= nums[i-1]
                ans.append(prefix)
        
        for i in range(len(nums)):
            if i == 0:
                continue
            else:
                
                postfix *= nums[-i]
                ans[-i-1] *= postfix
        return ans
                
            
            

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int n = nums.Length;
        int[] res = new int[n];
        
        // Step 1: Compute the left products and store them in res
        res[0] = 1;
        for (int i = 1; i < n; i++) {
            res[i] = res[i - 1] * nums[i - 1];
        }
        
        // Step 2: Compute the right products and update the res array
        int product = 1;
        for (int i = n - 2; i >= 0; i--) {
            product *= nums[i + 1];
            res[i] *= product;
        }
        
        return res;
    }
}
